mln/models/module.py

```diff
@@ -59,14 +60,96 @@ def _calc_yield_info(self):
+			remove_inv_item(clicker, cost.item_id, cost.qty)
+		guest_yield = self._get_guest_yield(self.item.guest_yields.all())
+		if guest_yield is not None: 
+			add_inv_item(clicker, guest_yield.item_id, guest_yield.qty)
+		for owner_yield in self.item.owner_yields.all():
+			if random.random() < (owner_yield.probability / 100): 
+				add_inv_item(self.owner, owner_yield.item_id, owner_yield.qty)
+		for friend_message in self.item.friend_messages.all():
+			if random.random() < (friend_message.probability / 100): 
+				random_friend = self._get_random_friend()
+				send_template(template=friend_message.message, sender=self.owner, recipient=random_friend)
+		return guest_yield
+
+	def _get_guest_yield(self, yields):
+		if len(yields) == 0: return None  # no yield for clicking
+		elif len(yields) == 1:  # standard, one-item yield
+			result = yields[0]
+			if random.random() < (result.probability / 100): return result
+			else: return None
+		else:  # lottery
+			random_number = random.randrange(100)
+			sum = 0
+			for prize in yields: 
+				sum += prize.probability
+				if sum > random_number: 
+					return prize
+			else: 
+				raise RuntimeError(f"Couldn't choose a guest yield for {self}")
+
+	def _get_info(self):
+		"""Get the ModuleInfo for this module."""
+		return ModuleInfo.objects.get(item_id=self.item_id)
+
+	def _get_random_friend(self): 
+		"""Returns a random friend from the module owner's friend list."""
+		# Wouldn't help to use Q objects because we wouldn't know which user is the friend
+		incoming = self.owner.incoming_friendships.filter(from_user__profile__is_networker=False, status=FriendshipStatus.FRIEND)
+		outgoing = self.owner.outgoing_friendships.filter(to_user__profile__is_networker=False, status=FriendshipStatus.FRIEND)
+		incoming_friends = [friendship.from_user for friendship in incoming]
+		outgoing_friends = [friendship.to_user for friendship in outgoing]
+		friends = incoming_friends + outgoing_friends
+		return random.choice(friends)
+
+	def _is_setupable(self):
+		return self._get_info().editor_type == ModuleEditorType.TRADE or ModuleSetupCost.objects.filter(module_item_id=self.item_id).exists()
+
+	def _update_clicks(self, clicker):
+		"""Updates clicks for both the clicker and the module."""
+		if clicker == self.owner:  # user can't click on own module
+			raise ValueError("Can't vote on own module")
+		elif clicker.profile.available_votes <= 0:  # user must have clicks
+			raise RuntimeError("Voter has no votes left")
+		elif not self.is_clickable():  # module must be set-up
+			raise RuntimeError("This module is not set up")
+		clicker.profile.available_votes -= 1
+		clicker.profile.save()
+		self.clicks_since_last_harvest += 1
+		self.total_clicks += 1
+		self.save()
+
+	def get_settings_classes(self):
+		"""Get the save data classes for this module."""
+		info = self._get_info()
+		if info.editor_type is None:
+			return ()
+		return module_settings_classes[info.editor_type]

```

Comment:
Maybe the dict should have None as a key with an empty list?

Reply:
I'd be fine with that as well, I leave it up to you.